Analysis/PIDTools/src/Compute_dEdxProcessor.cc

```diff
@@ -104,136 +177,305 @@ void Compute_dEdxProcessor::processEvent( LCEvent * evt ) {
+  dedx2 = dedx2/(float)(nTruncate2);
+  dedx3 = dedx3/(float)(nTruncate3);
+  streamlog_out(DEBUG4) << "  check dedx: " << dedx1 << " " << dedx2 << " " << dedx3 << std::endl;
+
+  double trkcos = sqrt(trk->getTanLambda()*trk->getTanLambda()/(1.0+trk->getTanLambda()*trk->getTanLambda()));
+  double normdedx1 = getNormalization(dedx1, (float)nTruncate1, trkcos);
+  double normdedx2 = getNormalization(dedx2, (float)nTruncate2, trkcos);
+  double normdedx3 = getNormalization(dedx3, (float)nTruncate3, trkcos);
+  if(_isSmearing)  //additional smearing 20170901
+  {
+    normdedx1 += getSmearing(normdedx1);
+    normdedx2 += getSmearing(normdedx2);
+    normdedx3 += getSmearing(normdedx3);
   }
 
-  delete [] depe;
+  if (_StratCompHist)
+  {
+    double lmom = sqrt(pow(trkHelixC.getMomentum()[0],2) + pow(trkHelixC.getMomentum()[1],2) + pow(trkHelixC.getMomentum()[2],2));
 
-  //cal. truncated mean
-  dedx = dedx/(float)(nTruncate);
-  //std::cout << "check dedx: " << dedx << " " << unum << std::endl;
+    if (_StratCompHistWeight) {
+      _BBHist_Strategy1->Fill(lmom,normdedx1,sqrt(nTruncate1));
+      _BBHist_Strategy2->Fill(lmom,normdedx2,sqrt(nTruncate2));
+      _BBHist_Strategy3->Fill(lmom,normdedx3,sqrt(nTruncate3));
+    }
+    else {
+      _BBHist_Strategy1->Fill(lmom,normdedx1);
+      _BBHist_Strategy2->Fill(lmom,normdedx2);
+      _BBHist_Strategy3->Fill(lmom,normdedx3);
+    }
+    if (nTruncate1>=30)
+    {
+      if (_StratCompHistWeight) {
+        _BBHist_n30_Strategy1->Fill(lmom,normdedx1,sqrt(nTruncate1));
+        _BBHist_n30_Strategy2->Fill(lmom,normdedx2,sqrt(nTruncate2));
+        _BBHist_n30_Strategy3->Fill(lmom,normdedx3,sqrt(nTruncate3));
+      }
+      else {
+        _BBHist_n30_Strategy1->Fill(lmom,normdedx1);
+        _BBHist_n30_Strategy2->Fill(lmom,normdedx2);
+        _BBHist_n30_Strategy3->Fill(lmom,normdedx3);
+      }
+    }
+  }
 
-  double trkcos=sqrt(trk->getTanLambda()*trk->getTanLambda()/(1.0+trk->getTanLambda()*trk->getTanLambda()));
-  float normdedx=getNormalization(dedx, (float) nTruncate, trkcos);
-  if(_isSmearing) normdedx += getSmearing(normdedx);   //additional smearing 20170901
+  double normdedx=0, totway=0;
+  switch (_dxStrategy)
+  {
+  case  1: normdedx = normdedx1; nTruncate = nTruncate1; totway = totlength; break;
+  case  2: normdedx = normdedx2; nTruncate = nTruncate2; totway = totpath;   break;
+  case  3: normdedx = normdedx3; nTruncate = nTruncate3; totway = totover;   break;
+  default: normdedx = normdedx1; nTruncate = nTruncate1; totway = totlength;
+  }
 
-  float *ret=new float[2];
-  ret[0]=normdedx;
-  ret[1]=normdedx*std::pow(0.001*totlength, -0.34)*std::pow(nTruncate, -0.45);   //todo: what is gas pressure in TPC?
+  std::pair<double,double> ret;
+  ret.first =normdedx;
+  ret.second=normdedx*std::pow(0.001*totway, -0.34)*std::pow(nTruncate, -0.45);   //todo: what is gas pressure in TPC?
 
-  if(ret[0]!=ret[0]){
-    ret[0]=0.0;
-    ret[1]=0.0;
+  if(ret.first!=ret.first){
```

Comment:
If you want to check for NAN as the next lines suggest, use `std::isnan(ret.first)` instead.

Reply:
done